### ISSUE TITLE
Ranch silently reduces accept rate when it runs out of fd's

### ISSUE THREAD

**Issue by mmzeeman** [ninenines/ranch#134]
When ranch runs out of fd's it silently reduces its request rate without producing a log or warning message. This leaves the owner unaware of a serious problem.

*Note: Reducing the request rate is a solution which can help in some situations. Another simple solution is crash. Then all open fd's are automatically dropped and the service can restart. This also gives some back-pressure to connected clients. Maybe the emfile handling should be configurable?*

**Comment by essen** [comment]
I'm not against adding a log message. Definitely should be documented in the user guide.

Crashing doesn't help, because it will crash a lot more than just the acceptor since it'll get restarted and likely will crash again. Crashing all connections because you run out of fds is not very good.

**Comment by mmzeeman** [comment]
Well, crashing should be avoided of course, but there are situations where dropping open connections works better than reducing the request rate. Clients should be able to deal with network related glitches anyway. 

I've tested dropping connections in practice with our current mochiweb acceptor in zotonic, which we are going to replace with ranch/cowboy in the near future, and counting the number of successful requests in overload situations. But indeed, YMMV.

I'll check if I can make a PR for the warning message for you.

BTW by looking at the code I also think there is a problem with respect to accepting ssl connections. It looks like ranch can run out of acceptors because the ssl handshake is done inside the accept loop instead of in the protocol handler. Is this a known issue? As you probably know, the handshake can take seconds to complete leaving the acceptor unavailable for new requests. This was something we fixed in our mochiweb branch because we noticed reduced ssl throughput.
